File: Rock.JavaScript.Obsidian/Framework/Utility/cache.ts
// <copyright>
// Copyright by the Spark Development Network
//
// Licensed under the Rock Community License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.rockrms.com/license
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>

import { RockDateTime } from "./rockDateTime";


type CacheEntry<T> = {
    value: T;
    expiration: number;
};

const keyPrefix = "ObsidianCache_";

/**
* Stores the value using the given key. The cache will expire at the expiration or in
* 1 minute if none is provided
* @param key Unique key to store the value under
* @param value Value to store
* @param ttl Time to live in seconds (how long the cached value is valid for). Defaults to 60 if not provided.
*/
function set<T>(key: string, value: T, ttl: number = 60): void {
    const expiration = RockDateTime.now().addSeconds(ttl).toMilliseconds();

    const cache: CacheEntry<T> = { expiration, value };
    const cacheJson = JSON.stringify(cache);
    sessionStorage.setItem(keyPrefix + key, cacheJson);
}

/**
 * Gets a stored cache value if there is one that has not yet expired.
 * @param key
 */
function get<T>(key: string): T | null {
    const cacheJson = sessionStorage.getItem(keyPrefix + key);

    if (!cacheJson) {
        return null;
    }

    const cache = JSON.parse(cacheJson) as CacheEntry<T>;

    if (!cache || !cache.expiration) {
        return null;
    }

    if (cache.expiration < RockDateTime.now().toMilliseconds()) {
        return null;
    }

    return cache.value;
}

// Stores promises that are in flight so that if a second fetch is fired before the first resolves
// (before value gets cached), just return the promise that's already in flight to prevent duplicate calls
const promiseCache: Record<string, Promise<unknown> | undefined> = {};

/**
 * Since Promises can't be cached, we need to store them in memory until we get the result back. This wraps
 * a function in another function that returns a promise and...
 * - If there's a cached result, return it
 * - Otherwise if there's a cached Promise, return it
 * - Otherwise call the given function and cache it's promise and return it. Once the the Promise resolves, cache its result
 *
 * @param key Key for identifying the cached values
 * @param fn Function that returns a Promise that we want to cache the value of
 * @param ttl Time to live in seconds (how long the cached value is valid for)
 *
 * @returns A function that wraps the given function, that can be called instead of the given function and that will use the cache when appropriate.
 */
function cachePromiseFactory<T>(key: string, fn: () => Promise<T>, ttl?: number): () => Promise<T> {
    return async function (): Promise<T> {
        // If it's cached, grab it
        const cachedResult = get<T>(key);
        if (cachedResult) {
            return cachedResult;
        }

        // If it's not cached yet but we've already started fetching it
        // (it's not cached until we receive the results), return the existing Promise
        if (promiseCache[key]) {
            return promiseCache[key] as Promise<T>;
        }

        // Not stored anywhere, so fetch it and save it on the stored Promise for the next call
        promiseCache[key] = fn();

        // Once it's resolved, cache the result
        promiseCache[key]?.then((result) => {
            set(key, result, ttl);
            delete promiseCache[key];
            return result;
        }).catch((e: Error) => {
            // Something's wrong, let's get rid of the stored promise, so we can try again.
            delete promiseCache[key];
            throw e;
        });

        return promiseCache[key] as Promise<T>;
    };
}


export default {
    set,
    get,
    cachePromiseFactory
};

File: Rock/Lava/RockLiquid/Blocks/Cache.cs
// <copyright>
// Copyright by the Spark Development Network
//
// Licensed under the Rock Community License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.rockrms.com/license
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>
//
using System;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;
using System.Runtime.Serialization;
using System.Text;
using System.Text.RegularExpressions;

using DotLiquid;
using DotLiquid.Util;

using Rock.Lava.Blocks;
using Rock.Lava.DotLiquid;
using Rock.Web.Cache;

namespace Rock.Lava.RockLiquid.Blocks
{
    /// <summary>
    /// Cache allows you to cache the results of a Lava template.
    ///
    /// {% cache key:'my-content' %}
    ///     My Lava is now fast!
    /// {% endcache %}
    /// </summary>
    public class Cache : RockLavaBlockBase
    {
        string _markup = string.Empty;
        string _tagName = string.Empty;

        StringBuilder _blockMarkup = new StringBuilder();

        /// <summary>
        /// Method that will be run at Rock startup
        /// </summary>
        public override void OnStartup()
        {
            Template.RegisterTag<Cache>( "cache" );
        }

        /// <summary>
        /// Initializes the specified tag name.
        /// </summary>
        /// <param name="tagName">Name of the tag.</param>
        /// <param name="markup">The markup.</param>
        /// <param name="tokens">The tokens.</param>
        /// <exception cref="System.Exception">Could not find the variable to place results in.</exception>
        public override void Initialize( string tagName, string markup, List<string> tokens )
        {
            _markup = markup;
            _tagName = tagName;

            base.Initialize( tagName, markup, tokens );
        }

        /// <summary>
        /// Parses the specified tokens.
        /// </summary>
        /// <param name="tokens">The tokens.</param>
        protected override void Parse( List<string> tokens )
        {
            // Get the block markup. The list of tokens contains all of the lava from the start tag to
            // the end of the template. This will pull out just the internals of the block.

            // We must take into consideration nested tags of the same type

            // This is similar logic to the Shortcodes, but the tag regex are different. Attempted to refactor to a reusable helper, but it needs
            // access to a lot of the internals of the command.

            var endTagFound = false;

            var startTag = $@"{{\%\s*{ _tagName }\s*\%}}";
            var endTag = $@"{{\%\s*end{ _tagName }\s*\%}}";

            var childTags = 0;

            Regex regExStart = new Regex( startTag );
            Regex regExEnd = new Regex( endTag );

            NodeList = NodeList ?? new List<object>();
            NodeList.Clear();

            string token;
            while ( ( token = tokens.Shift() ) != null )
            {

                Match startTagMatch = regExStart.Match( token );
                if ( startTagMatch.Success )
                {
                    childTags++; // increment the child tag counter
                    _blockMarkup.Append( token );
                }
                else
                {
                    Match endTagMatch = regExEnd.Match( token );

                    if ( endTagMatch.Success )
                    {
                        if ( childTags > 0 )
                        {
                            childTags--; // decrement the child tag counter
                            _blockMarkup.Append( token );
                        }
                        else
                        {
                            endTagFound = true;
                            break;
                        }
                    }
                    else
                    {
                        _blockMarkup.Append( token );
                    }
                }
            }

            if ( !endTagFound )
            {
                AssertMissingDelimitation();
            }
        }

        /// <summary>
        /// Renders the specified context.
        /// </summary>
        /// <param name="context">The context.</param>
        /// <param name="result">The result.</param>
        public override void Render( Context context, TextWriter result )
        {
            // First ensure that cached commands are allowed in the context
            if ( !this.IsAuthorized( context ) )
            {
                result.Write( string.Format( RockLavaBlockBase.NotAuthorizedMessage, this.Name ) );
                base.Render( context, result );
                return;
            }

            var settings = LavaElementAttributes.NewFromMarkup( _markup, new RockLiquidRenderContext( context ) );

            var twoPassEnabled = settings.GetBoolean( "twopass", false );

            // Make sure the cache key is unique to this Lava engine.
            var cacheKey = "lavacache-dotliquid-" + settings.GetString( "key", string.Empty );
            if ( cacheKey == string.Empty )
            {
                result.Write( "* No cache key provided. *" );
                base.Render( context, result );
                return;
            }

            // Get content from cache
            var cachedResult = RockCache.Get( cacheKey, true ) as CacheLavaTag;

            // Check that the cached value is current
            if ( cachedResult != null )
            {
                var currentHash = CalculateContentHash( _blockMarkup.ToString() );
                if ( currentHash != cachedResult.Hash )
                {
                    cachedResult = null;
                }
            }

            // Use the cached value
            if ( cachedResult != null )
            {
                if ( twoPassEnabled )
                {
                    result.Write( MergeLava( cachedResult.Content, context ) );
                }
                else
                {
                    result.Write( cachedResult.Content );
                }

                base.Render( context, result );
                return;
            }

            // Cached value not available so render the template and cache it
            var lavaResults = MergeLava( _blockMarkup.ToString(), context );

            var cacheDuration = settings.GetInteger( "duration", 3600 );

            if ( cacheDuration > 0 )
            {
                // Don't cache if it's too large
                var maxCacheSize = settings.GetInteger( "maxcachesize", 200000 );

                if ( lavaResults.Length < maxCacheSize )
                {
                    var expiration = RockDateTime.Now.AddSeconds( cacheDuration );
                    var cachedHash = CalculateContentHash( _blockMarkup.ToString() );
                    RockCache.AddOrUpdate( cacheKey, string.Empty,
                        new CacheLavaTag { Hash = cachedHash, Content = lavaResults },
                        expiration,
                        settings.GetString( "tags", string.Empty ) );
                }
            }

            // If twopass is enabled run the lava again
            if ( twoPassEnabled )
            {
                lavaResults = MergeLava( lavaResults, context );
            }

            result.Write( lavaResults );



            base.Render( context, result );
        }

        /// <summary>
        /// Calculates the content hash.
        /// </summary>
        /// <param name="content">The content.</param>
        /// <returns></returns>
        private int CalculateContentHash( string content )
        {
            return ( content + _markup ).GetHashCode();
        }

        /// <summary>
        /// Cache Tag POCO
        /// </summary>
		[Serializable]
        [DataContract]
        private class CacheLavaTag
        {
            /// <summary>
            /// Gets or sets the hash.
            /// </summary>
            /// <value>
            /// The hash.
            /// </value>
			[DataMember]
            public int Hash { get; set; }

            /// <summary>
            /// Gets or sets the lava.
            /// </summary>
            /// <value>
            /// The lava.
            /// </value>
			[DataMember]
            public string Content { get; set; }
        }

        /// <summary>
        /// Merges the lava.
        /// </summary>
        /// <param name="lavaTemplate">The lava template.</param>
        /// <param name="context">The context.</param>
        /// <returns></returns>
        private string MergeLava( string lavaTemplate, Context context )
        {
            // Get enabled commands
            var enabledCommands = context.Registers["EnabledCommands"].ToString();

            // Get mergefields from lava context
            var lavaMergeFields = new Dictionary<string, object>();
            if ( context.Environments?.Count > 0 )
            {
                foreach ( var item in context.Environments[0] )
                {
                    lavaMergeFields.Add( item.Key, item.Value );
                }
            }

            // Add variables in the scope (defined in the lava itself via assign)
            if ( context.Scopes?.Count > 0 )
            {
                foreach ( var scope in context.Scopes )
                {
                    foreach ( var item in scope )
                    {
                        lavaMergeFields.Add( item.Key, item.Value );
                    }
                }
            }

            return lavaTemplate.ResolveMergeFields( lavaMergeFields, enabledCommands );
        }
    }
}